### PR TITLE
BUG: raise on non-hashable in __contains__

### PR DESCRIPTION
**Pull request by jbrockmendel**
By checking before we get to the `Engine.__contains__` call, we can avoid a redundant call to `hash`

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]
lgtm

File: pandas/_libs/index.pyx [hunk 1]
```diff
@@ -72,9 +72,10 @@ cdef class IndexEngine:
         self.over_size_threshold = n >= _SIZE_CUTOFF
         self.clear_mapping()
 
-    def __contains__(self, object val):
+    def __contains__(self, val: object) -> bool:
+        # We assume before we get here:
```
Comment (simonjayhawkins) [inline pandas/_libs/index.pyx]: can val be typed as Hashable, (comment then not needed?)

File: pandas/_libs/index.pyx [hunk 2]
```diff
@@ -409,7 +408,9 @@ cdef class DatetimeEngine(Int64Engine):
     cdef _get_box_dtype(self):
         return 'M8[ns]'
 
-    def __contains__(self, object val):
+    def __contains__(self, val: object) -> bool:
```
Comment (simonjayhawkins) [inline pandas/_libs/index.pyx]: same

File: pandas/_libs/index.pyx [hunk 3]
```diff
@@ -717,7 +718,9 @@ cdef class BaseMultiIndexCodesEngine:
 
         return indexer
 
-    def __contains__(self, object val):
+    def __contains__(self, val: object) -> bool:
```
Comment (simonjayhawkins) [inline pandas/_libs/index.pyx]: same

File: pandas/core/indexes/category.py [hunk 4]
```diff
@@ -390,6 +390,7 @@ def __contains__(self, key) -> bool:
         if is_scalar(key) and isna(key):
             return self.hasnans
 
+        hash(key)
```
Comment (simonjayhawkins) [inline pandas/core/indexes/category.py]: I assume this is to check that key is hashable. can you not type key as Hashable?

File: pandas/tests/indexes/common.py [hunk 5]
```diff
@@ -883,3 +883,11 @@ def test_getitem_2d_deprecated(self):
             res = idx[:, None]
 
         assert isinstance(res, np.ndarray), type(res)
+
+    def test_contains_requires_hashable(self):
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/common.py]:
```suggestion
    def test_contains_requires_hashable_raises(self):
```

File: pandas/tests/indexes/common.py [hunk 6]
```diff
@@ -883,3 +883,11 @@ def test_getitem_2d_deprecated(self):
...
+        idx = self.create_index()
+        with pytest.raises(TypeError, match="unhashable type"):
+            [] in idx
+
+        with pytest.raises(TypeError):
+            {} in idx._engine
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/common.py]: do we/should we have a non-hashable fixture?

File: pandas/tests/indexes/common.py [hunk 7]
```diff
@@ -883,3 +883,11 @@ def test_getitem_2d_deprecated(self):
             res = idx[:, None]
 
         assert isinstance(res, np.ndarray), type(res)
+
+    def test_contains_requires_hashable(self):
+        idx = self.create_index()
+        with pytest.raises(TypeError, match="unhashable type"):
+            [] in idx
+
+        with pytest.raises(TypeError):
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/common.py]: missing match argument

**Review by simonjayhawkins — COMMENTED** [review]
@jbrockmendel the issue title includes BUG, is there a related issue?

File: pandas/_libs/index.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/index.pyx]: im planning to add more assumptions to the comment in future PRs

**Comment by jbrockmendel** [comment]
mypy didnt like my attempt at annotating, so reverted

**Comment by simonjayhawkins** [comment]
> mypy didnt like my attempt at annotating, so reverted

is this because \_\_contains__ is a special method and this code is unsafe. Is this not also a breaking change?

**Comment by jbrockmendel** [comment]
restored annotations.  turns out the trouble was that `RangeIndex.__contains__` had a non-matching annotation

**Comment by simonjayhawkins** [comment]
> restored annotations. turns out the trouble was that `RangeIndex.__contains__` had a non-matching annotation

I think mypy should only complain if the class is a subclass of collections.abc.Container
even so, in typeshed, classes that want to restrict the type of key add a type:ignore.

so rather than Any, can Hashable be used?

**Comment by jbrockmendel** [comment]
> so rather than Any, can Hashable be used?

When I use `Optional[Hashable]` mypy gives:

```
pandas/core/indexes/numeric.py:236: error: No overload variant of "int" matches argument type "Optional[Hashable]"
pandas/core/indexes/numeric.py:236: note: Possible overload variant:
pandas/core/indexes/numeric.py:236: note:     def int(self, x: Union[str, bytes, SupportsInt] = ...) -> int
pandas/core/indexes/numeric.py:236: note:     <1 more non-matching overload not shown>
pandas/core/indexes/numeric.py:488: error: Argument 1 to "len" has incompatible type "Optional[Hashable]"; expected "Sized"
pandas/core/indexes/numeric.py:488: error: Item "Hashable" of "Optional[Hashable]" has no attribute "item"
pandas/core/indexes/numeric.py:488: error: Item "None" of "Optional[Hashable]" has no attribute "item"
pandas/core/indexes/numeric.py:490: error: Argument 1 to "len" has incompatible type "Optional[Hashable]"; expected "Sized"
Found 5 errors in 1 file (checked 846 source files)
```

**Comment by WillAyd** [comment]
Yea I guess might be tough to get that to work with mypy since we are catching the `TypeError` at runtime, though passing a non-hashable item to that would currently raise an error at the `hash()` call that is uncaught, so I think mypy is highlighting an actual bug in a roundabout way

File: pandas/_libs/index.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/index.pyx]: you *could* add an assert

File: pandas/core/indexes/category.py (hunk 4 above)
Comment (jreback) [reply pandas/core/indexes/category.py]: that's not a run-time check

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine. can you add assert hash(key) in the cython code?

**Comment by jbrockmendel** [comment]
> looks fine. can you add assert hash(key) in the cython code?

A big part of the motivation here is to avoid duplicated hash calls

**Comment by jreback** [comment]
> > looks fine. can you add assert hash(key) in the cython code?
> 
> A big part of the motivation here is to avoid duplicated hash calls

do asserts not get stripped? isn't that the point? (its not a big deal as long as you have a comment, but the asserts are to provide developer guarantees)

also needs a rebase

**Review by jorisvandenbossche — APPROVED** [review]
Nice clean-up!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**